ipld/get_namespaced_shares.go

```diff
@@ -0,0 +1,185 @@
+package ipld
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/gammazero/workerpool"
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+
+	"github.com/celestiaorg/celestia-node/ipld/plugin"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+// TODO(@distractedm1nd) Find a better figure than NumWorkersLimit for this pool.
+// Worker pool responsible for the goroutines spawned by getLeavesByNamespace
+var namespacePool = workerpool.New(NumWorkersLimit)
+
+// GetSharesByNamespace walks the tree of a given root and returns its shares within the given namespace.ID.
+func GetSharesByNamespace(
+	ctx context.Context,
+	bGetter blockservice.BlockGetter,
+	root cid.Cid,
+	nID namespace.ID,
+	maxShares int,
+) ([]Share, error) {
+	leaves, err := getLeavesByNamespace(ctx, bGetter, root, nID, maxShares)
+	if err != nil {
+		return nil, err
+	}
+
+	shares := make([]Share, len(leaves))
+	for i, leaf := range leaves {
+		shares[i] = leafToShare(leaf)
+	}
+
+	return shares, nil
+}
+
+// wrappedWaitGroup is needed because waitgroups do not expose their internal counter,
+// and we don't know in advance how many jobs we will have to wait for.
+type wrappedWaitGroup struct {
+	wg      sync.WaitGroup
+	counter int64
+}
+
+func (w *wrappedWaitGroup) Add(count int64) {
+	w.wg.Add(int(count))
+	atomic.AddInt64(&w.counter, count)
+}
+
+func (w *wrappedWaitGroup) Done(jobs chan *job) {
+	w.wg.Done()
+	atomic.AddInt64(&w.counter, -1)
+	if atomic.LoadInt64(&w.counter) == 0 {
+		close(jobs)
+	}
+}
+
+func (w *wrappedWaitGroup) Wait() {
+	w.wg.Wait()
+}
+
+func (w *wrappedWaitGroup) IsWorking() bool {
+	return atomic.LoadInt64(&w.counter) > 0
+}
+
+type fetchedBounds struct {
+	lowest  int64
+	highest int64
+}
+
+func (b *fetchedBounds) Update(index int64) bool {
+	ok := true
+	if lowest := atomic.LoadInt64(&b.lowest); index < lowest {
+		ok = ok && atomic.CompareAndSwapInt64(&b.lowest, lowest, index)
+	}
+	// not an `else if` because an element can be both the lower and higher bound
+	// for example, if there is only one share in the namespace
+	if highest := atomic.LoadInt64(&b.highest); index > highest {
+		ok = ok && atomic.CompareAndSwapInt64(&b.highest, highest, index)
+	}
+	return ok
+}
```

Comment:
nit: I totally see the code cleanness idea here, but it would be even better from a struct API standpoint to not have to return a bool and make the use of it to loop it until it's true. 
* Two `for`s inside and making `Update` atomically guarantee the success of the update operation > making the user worry about calling for over it.
* Also we avoid idle `lowest` loadings, when `highest` failed to be swapped in the first attempt